File: PracticasPDOOJava/Practica3_Java/src/irrgarten/Labyrinth.java
package irrgarten;

import java.util.ArrayList;
/**
 * Representa el laberinto del juego como tres tablas paralelas:
 * <ul>
 *          <li> {@code monsters[r][c]}:    referencia al monstruo en la celda (r, c) o {@code null}. </li>  
 *          <li> {@code players[r][c]}:            referencia al jugador en la celda (r, c) o {@code null}. </li>
 *          <li> {@code cells[r][c]}:                 carácter con el estado visual de la celda: 
 *                  'X' (obstáculo), '-' (vacía), 'M' (monstruo)
 *                  'E' (salida), 'C' (combat: jugador y monstruo) </li>
 * </ul>
 * @author Joaquín Cruz Lorenzo
 */
public class Labyrinth {
    // ===============================
    // CONSTANTES DE CLASE
    // ===============================
    private static final char BLOCK_CHAR = 'X';
    private static final char EMPTY_CHAR = '-';
    private static final char MONSTER_CHAR = 'M';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    public static final int ROW = 0;
    public static final int COL = 1;
    
    // ===============================
    // ATRIBUTOS DE INSTANCIA
    // ===============================
    private int nRows, nCols;
    private int exitRow, exitCol;
    
    private final Monster[][] monsters;
    private final Player[][] players;
    private final char[][] labyrinth;
    
    // ===============================
    // MÉTODOS PÚBLICOS
    // ===============================

    /**
     * Crea el laberinto e inicializa todas las celdas como vacías ('-'),
     * fijando la casilla de salida en {@code (exitRow, exitCol)} a 'E'.
     * 
     * @param nRows      número de filas
     * @param nCols       número de columnas
     * @param exitRow   fila de la salida
     * @param exitCol    columna de la salida
     */
    public Labyrinth(int nRows, int nCols) {
        this.nRows = nRows;
        this.nCols = nCols;
        
        // Creamos los arrays bidimensionales
        this.monsters = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.labyrinth = new char[nRows][nCols];
        
        // 1. Inicializar todo a EMPTY_CHAR ('-')
        for (int i = 0; i < nRows; i++)
            for (int j = 0; j < nCols; j++){
                this.labyrinth[i][j] = EMPTY_CHAR;
                this.monsters[i][j] = null; // Aseguro que empiezan vacíos
                this.players[i][j] = null; // Aseguro que empiezan vacíos
            }
        
        // 2. ANIADO MUROS EXTERIORES (BLOCK_CHAR = 'X')
        // Filas superior e inferior
        for (int j = 0; j < nCols; j++) {
            this.labyrinth[0][j] = BLOCK_CHAR; // Primera fila
            this.labyrinth[nRows - 1][j] = BLOCK_CHAR; // Última fila
        }

        // Columnas izquierda y derecha (sin sobreescribir)
        for (int i = 1; i < nRows - 1; i++){
            this.labyrinth[i][0] = BLOCK_CHAR; // Primera columna
            this.labyrinth[i][nCols - 1] = BLOCK_CHAR; // Última columna
        }
        
        // 3. ANIADIR MUROS INTERIORES 
        if (posOK(2, 2) && !(exitRow == 2 && exitCol == 2)) this.labyrinth[2][2] = BLOCK_CHAR;
        if (posOK(2, 3) && !(exitRow == 2 && exitCol == 3)) this.labyrinth[2][3] = BLOCK_CHAR;
        if (posOK(3, 2) && !(exitRow == 3 && exitCol == 2)) this.labyrinth[3][2] = BLOCK_CHAR;
        
        // 4. GENERAR Y COLOCAR LA SALIDA ALEATORIA INTERNA 
        // Genera coordenadas aleatorias hasta encontrar una casilla interna que NO sea un muro ('X')
        int randomExitRow, randomExitCol;
        
        do {
            // Genera fila entre 1 y nRows - 2 (inclusive)
            randomExitRow = Dice.randomPos(nRows - 2) + 1;
            // Genera columna entre 1 y nCols - 2 (inclusive)
            randomExitCol = Dice.randomPos(nCols - 2) + 1;
        } while (this.labyrinth[randomExitRow][randomExitCol] == BLOCK_CHAR); // Repite si cae en un muro interno
        
        // Asignar y colocar la salida
        this.exitRow = randomExitRow; // Guarda la fila generada
        this.exitCol = randomExitCol; // Guarda la columna generada
        this.labyrinth[this.exitRow][this.exitCol] = EXIT_CHAR; // Poner 'E'
    }
    
    /**
     * @return {@code true} si hay un jugador en la casilla de salida.
     */
    public boolean haveAWinner() {
        return players[exitRow][exitCol] != null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        
        for(int i = 0; i < nRows; i++) {
            for(int j = 0; j < nCols; j++)
                sb.append(labyrinth[i][j]).append("\t");
            
            sb.append("\n");
        }
        
        return sb.toString();
    }
 
    /**
     * Añade un monstruo si la posición es válida y está vacía.
     * Marca la celda con 'M' y fija su posición.
     * 
     * @param row           fila destino
     * @param col            columna destino
     * @param monster  monstruo a colocar
     */
    public void addMonster(int row, int col, Monster monster) {
        if (posOK(row, col) && emptyPos(row, col)) {
            monsters[row][col] = monster;
            labyrinth[row][col] = MONSTER_CHAR;
            monster.setPos(row, col);
        }
    }
    
    /**
     * Este método se mantiene público, desviándose del UML original, por decisión de disenio
     * para permitir una configuración aleatoria del tablero desde la clase Game. Esta decisión 
     * fue consultada y aprobada por la profesora.
     * @return Array de dos enteros con la fila y columna de una casilla vacía.
     */
    public int[] randomEmptyPos() {
        while(true) {
            int r = Dice.randomPos(nRows);
            int c = Dice.randomPos(nCols);
            
            if(emptyPos(r, c)) return new int[]{r, c};
        }
    }
    
    // ==============================
    // P3 - Métodos públicos a implementar
    // ==============================
    /**
     * Devuelve una lista de direcciones válidas desde una casilla.
     * @param row
     * @param col
     * @return Lista de direcciones a las que se puede mover
     */
    public ArrayList<Directions> validMoves(int row, int col) { 
        ArrayList<Directions> output = new ArrayList<>(); // Diagrama 1.1: new
        
        // Diagrama opt: [(canStepOn(row + 1, col))]
        if (canStepOn(row + 1, col)) {
            output.add(Directions.DOWN); // Diagrama 1.2: add(Directions.DOWN)
        }
        
        // Diagrama opt: [(canStepOn(row - 1, col))]
        if (canStepOn(row - 1, col)) {
            output.add(Directions.UP); // Diagrama 1.3: add(Directions.UP)
        }
        
        // Diagrama opt: [(canStepOn(row, col + 1))]
        if (canStepOn(row, col + 1)) {
            output.add(Directions.RIGHT); // Diagrama 1.4: add(Directions.RIGHT)
        }
        
        // Diagrama opt: [(canStepOn(row, col - 1))]
        if (canStepOn(row, col - 1)) {
            output.add(Directions.LEFT); // Diagrama 1.5: add(Directions.LEFT)
        }
        
        return output; // Diagrama 1.6: output
    }
    
    /**
     * Aniade un bloque de obstáculos (muros) al laberinto
     * @param orientation
     * @param startRow
     * @param startCol
     * @param length 
     */
    public void addBlock(Orientation orientation, int startRow, int startCol, int length) { 
        int incRow, incCol;
        
        // Diagrama alt: [(orientation == Orientation.VERTICAL)]
        if (orientation == Orientation.VERTICAL) {
            incRow = 1;
            incCol = 0;
        } else { // HORIZONTAL
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        // Diagrama loop: [( (posOK(row, col)) && (emptyPos(row, col)) && (length > 0)
        while (posOK(row, col) && emptyPos(row, col) && length > 0) {
            // Diagrama 1.1: set(row, col, BLOCK_CHAR)
            labyrinth[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }
    
    /**
     * Mueve un jugador en una dirección
     * @param direction
     * @param player
     * @return El monstruo en la nueva casilla si hay, o null
     */
    public Monster putPlayer(Directions direction, Player player) { 
        int oldRow = player.getRow(); // Diagrama 1.1: getRow()
        int oldCol = player.getCol(); // Diagrama 1.2: getCol()
        
        int[] newPos = dir2Pos(oldRow, oldCol, direction); // Diagrama 1.3: dir2Pos(oldRow, oldCol, direction)
        
        Monster monster = putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player); // Diagrama 1.4: putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player)
        
        return monster; // Diagrama 2: monster
    }
    
    /**
     * Coloca a los jugadores en casullas vacías aleatorias.
     * @param players 
     */
    public void spreadPlayers(ArrayList<Player> players) { 
        // Diagrama loop: [for all]
        for (Player p : players) { // Diagrama 1.1: next()
            int[] pos = randomEmptyPos(); // Diagrama 1.2: randomEmptyPos()
            
            putPlayer2D(-1, -1, pos[ROW], pos[COL], p); // Diagrama 1.3: putPlayer2D(-1, -1, pos[ROW], pos[COL], p)
        }
    }
    
    
    
    // ============================
    // MÉTODOS PRIVADOS
    // ============================
    
    /**
     * 
     * @param row
     * @param col
     * @return {@code true} si (row, col) está dentro del tablero.
     */
    private boolean posOK(int row, int col) {
        return row >= 0 && row < nRows && col >= 0 && col < nCols;
    }
    
    /**
     * 
     * @param row
     * @param col
     * @return {@code true} si (row, col) es vacío
     */
    private boolean emptyPos(int row, int col) {
        return labyrinth[row][col] == EMPTY_CHAR;
    }
    
    /**
     * 
     * @param row
     * @param col
     * @return {@code true} si (row, col) es un monstruo
     */
    private boolean monsterPos(int row, int col) {
        return labyrinth[row][col] == MONSTER_CHAR;
    }
    
    /**
     * 
     * @param row
     * @param col
     * @return {@code true} si (row, col) está en exit
     */
    private boolean exitPos(int row, int col) {
        return labyrinth[row][col] == EXIT_CHAR;
    }
    
    /**
     * 
     * @param row
     * @param col
     * @return {@code true} si (row, col) es combate
     */
    private boolean combatPos(int row, int col) {
        return labyrinth[row][col] == COMBAT_CHAR;
    }
    
    /**
     * Indica si se puede pisar la celda: debe estar dentro del tablero y ser vacía
     * de monstruo o de salida (no muros).
     * @param row
     * @param col
     * @return 
     */
    private boolean canStepOn(int row, int col) {
        return posOK(row, col) 
                    && (emptyPos(row, col) || monsterPos(row, col) || exitPos(row, col));
    }
    
    /**
     * Deja la celda abandonada en estado correcto.
     * Si estaba en combate ('C') pasa a 'M', en caso contrario pasa a '-'.
     * Solo actúa si la posición es válida.ñ
     * @param row
     * @param col 
     */
    private void updateOldPos(int row, int col) {
        if(!posOK(row, col)) return;
        
        if(labyrinth[row][col] == COMBAT_CHAR) labyrinth[row][col] = MONSTER_CHAR;
        else labyrinth[row][col] = EMPTY_CHAR;
        
        players[row][col] = null; // Por coherencia, ninguna referencia del jugador
    } 
    
    /**
     * Calcula la posición a la que se llegaría desde (row, col) avanzando una 
     * casilla en la {@code direction} indicada. No comprueba límites.
     * @param row
     * @param col
     * @param direction
     * @return par [row, col] resultante.
     */
    private int[] dir2Pos(int row, int col, Directions direction) {
        switch(direction) {
            case LEFT : return new int[]{row, col - 1};
            case RIGHT : return new int[]{row, col + 1};
            case UP : return new int[]{row - 1, col};
            case DOWN : return new int[]{row + 1, col};
            default : return new int[]{row, col};
        }
    }
    
    // ===========================================
    // P3 - Métodos privados a implementar
    // ===========================================
    /**
     * Núcleo de la lógica del movimiento. Mueve un jugador de old a new
     * @param oldRow
     * @param oldCol
     * @param row
     * @param col
     * @param player
     * @return 
     */
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player) { 
        Monster output = null; // Diagrama: output = null
        
        // Diagrama opt: [canStepOn(row, col)]
        if (canStepOn(row, col)) {
            // Diagrama opt: [posOK(oldRow, oldCol)]
            if (posOK(oldRow, oldCol)) {
                Player p = players[oldRow][oldCol]; // Diagrama 1.1: get(oldRow, oldCol)
                
                // Diagrama opt: [p == player]
                if (p == player) {
                    updateOldPos(oldRow, oldCol); // Diagrama 1.2: updateOldPos(oldRow, oldCol)
                    players[oldRow][oldCol] = null; // Diagrama 1.3: set(oldRow, oldCol, null)
                }
            }
            
            boolean monsterPos = monsterPos(row, col); // Diagrama 1.4: monsterPos(row, col)
            
            // Diagrama alt: [monsterPos]
            if (monsterPos) {
                labyrinth[row][col] = COMBAT_CHAR; // Diagrama 1.5: set(row, col, COMBAT_CHAR)
                
                output = monsters[row][col]; // Diagrama 1.6: get(row, col)
            } else {
                char number = player.getNumber(); // Diagrama 1.7: getNumber()
                labyrinth[row][col] = number; // Diagrama 1.8: set(row, col, number)
            }
            
            players[row][col] = player; // Diagrama 1.9: set(row, col, player)
            
            player.setPos(row, col); // Diagrama 1.10: setPos(row, col)
        }
        
        return output; // Diagrama: output
    }
}

File: PracticasPDOORuby/Practica3_Ruby/lib/Irrgarten/labyrinth.rb
# frozen_string_literal: true
require_relative 'directions'
require_relative 'orientation'
require_relative "dice"

module Irrgarten
  class Labyrinth
    # Atributos de clase
    BLOCK_CHAR = 'X'
    EMPTY_CHAR = '-'
    MONSTER_CHAR = 'M'
    COMBAT_CHAR = 'C'
    EXIT_CHAR = 'E'
    ROW = 0
    COL = 1

    # --- Métodos públicos ---
    # Constructor
    # @param n_rows Número de filas del laberinto
    # @param n_cols Número de columnas del laberinto
    # @param exit_row Fila de la salida del laberinto
    # @param exit_col Columna de la salida del laberinto
    # @return Objeto de la clase Labyrinth
    def initialize(n_rows, n_cols, exit_row, exit_col)
      @n_rows = n_rows
      @n_cols = n_cols
      @exit_row = exit_row
      @exit_col = exit_col

      @monsters = Array.new(n_rows) { Array.new(n_cols) }
      @players = Array.new(n_rows) { Array.new(n_cols) }
      @labyrinth = Array.new(n_rows) { Array.new(n_cols, EMPTY_CHAR) }
      @labyrinth[exit_row][exit_col] = EXIT_CHAR
    end

    # Indica si hay un ganador (un jugador en la salida)
    # @return true si hay un ganador, false en caso contrario
    def have_a_winner?
      !@players[@exit_row][@exit_col].nil?
    end

    # Representación en String del laberinto
    # @return String con la representación del laberinto
    def to_s
      @labyrinth.map { |row| row.join("\t") }.join("\n")
    end

    # Añade un monstruo en una posición concreta del laberinto
    # @param row Fila donde se quiere añadir el monstruo
    # @param col Columna donde se quiere añadir el monstruo
    # @param monster Objeto de la clase Monster a añadir
    # @return Nada (modifica el estado del laberinto)
    def add_monster(row, col, monster)
      if pos_ok?(row, col) && empty_pos?(row, col)
        @labyrinth[row][col] = MONSTER_CHAR
        @monsters[row][col] = monster
        monster.set_pos(row, col)
      end
    end

    # Genera una posición vacía aleatoria en el laberinto
    # @return Array con la fila y columna de una posición vacía aleatoria
    def random_empty_pos
      loop do
        row = Dice.random_pos(@n_rows)
        col = Dice.random_pos(@n_cols)
        return [row, col] if empty_pos?(row, col)
      end
    end

    # ======================================================
    # P3 - Métodos públicos a implementar
    # ======================================================

    # Coloca al jugador / los jugadores en casilla/s vacía/s aleatoria/s.
    # @param players
    def spread_player(players)
      # Diagrama loop: [for all]
      players.each do |p| # Diagrama 1.1: next()
        random_empty_pos # Diagrama 1.2: random_empty_pos()
        put_player_2d(-1, -1, pos[ROW], pos[COL], p) # Diagrama 1.3: put_player_2d(-1, -1, row, col, player)
      end
    end

    # Mueve un jugador en una dirección.
    # @param direction
    # @param player
    # @return El monstruo en la nueva casilla, o nil
    def put_player(direction, player)
      old_row = player.row # Diagrama 1.1: getRow()
      old_col = player.col # Diagrama 1.2: getCol()

      new_pos = dir_2_pos(old_row, old_col, direction) # Diagrama 1.3: dir2Pos(oldRow, oldCol, direction)

      monster = put_player_2d(old_row, old_col, new_pos[ROW], new_pos[COL], player) # Diagrama 1.4: putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player)

      monster # Diagrama 2: return monster
    end

    # Aniade un bloque de obstáculos (muros al laberinto).
    # @param orientation
    # @param start_row
    # @param start_col
    # @param length
    def add_block(orientation, start_row, start_col, length)
      # Diagrama alt: [(orientation == Orientation::VERTICAL)]
      if orientation == Orientation::VERTICAL
        inc_row = 1
        inc_col = 0
      else
        inc_row = 0
        inc_col = 1
      end

      row = start_row
      col = start_col

      # Diagrama loop: [( (posOK(row, col)) && (emptyPos(row, col)) && (length > 0) )]
      while pos_ok?(row, col) && empty_pos?(row, col) && length > 0
        @labyrinth[row][col] = BLOCK_CHAR # Diagrama 1.1: set(row, col, BLOCK_CHAR)

        length -= 1
        row += inc_row
        col += inc_col
      end
    end

    # Devuelve un array de direcciones válidas desde una casilla.
    # @param row
    # @param col
    # @return Array de direcciones a las que se puede mover
    def valid_moves(row, col)
      output = [] # Diagrama 1.1: new

      # Diagrama opt: [canStepOn?(row + 1, col)]
      if can_step_on?(row + 1, col)
        output << Directions::DOWN # Diagrama 1.2: add(Directions::DOWN)
      end

      # Diagrama opt: [canStepOn?(row - 1, col)]
      if can_step_on?(row -1, col)
        output << Directions::UP # Diagrama 1.3: add(Directions::UP)
      end

      # Diagrama opt: [canStepOn?(row, col + 1)]
      if can_step_on?(row, col + 1)
        output << Directions::RIGHT # Diagrama 1.4: add(Directions::RIGHT)
      end

      # Diagrama opt: [canStepOn?(row, col -1)]
      if can_step_on?(row, col -1)
        output << Directions::LEFT # Diagrama 1.5: add(Directions::LEFT)
      end

      output # Diagrama 1.6: return output
    end


    # --- Métodos privados ---
    # Los siguientes métodos son de uso interno de la clase Labyrinth
    private

    # Comprueba si una posición está dentro de los límites del laberinto
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si la posición es válida, false en caso contrario
    def pos_ok?(row, col)
      row.between?(0, @n_rows - 1) && col.between?(0, @n_cols - 1)
    end

    # Comprueba si una posición está vacía
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si la posición está vacía, false en caso contrario
    def empty_pos?(row, col)
      pos_ok?(row, col) && @labyrinth[row][col] == EMPTY_CHAR
    end

    # Comprueba si una posición tiene un monstruo
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si la posición tiene un monstruo, false en caso contrario
    def monster_pos(row, col)
      pos_ok?(row, col) && @labyrinth[row][col] == MONSTER_CHAR
    end

    # Comprueba si una posición es la salida
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si la posición es la salida, false en caso contrario
    def exit_pos?(row, col)
      pos_ok?(row, col) && @labyrinth[row][col] == EXIT_CHAR
    end

    # Comprueba si una posición es de combate (monstruo debilitado)
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si la posición es de combate, false en caso contrario
    def combat_pos?(row, col)
      pos_ok?(row, col) && @labyrinth[row][col] == COMBAT_CHAR
    end

    # Comprueba si se puede pisar una posición (vacía, de combate o la salida)
    # @param row Fila de la posición a comprobar
    # @param col Columna de la posición a comprobar
    # @return true si se puede pisar la posición, false en caso contrario
    def can_step_on?(row, col)
      pos_ok?(row, col) && empty_pos?(row, col) || combat_pos?(row, col) || exit_pos?(row, col)
    end

    # Actualiza la posición antigua del jugador en el laberinto
    # @param row Fila de la posición antigua
    # @param col Columna de la posición antigua
    # @return Nada (modifica el estado del laberinto)
    def update_old_pos(row, col)
      if pos_ok?(row, col)
        @labyrinth[row][col] = combat_pos?(row, col) ? MONSTER_CHAR : EXIT_CHAR
      end
    end

    # Calcula la nueva posición a partir de una dirección
    # @param row Fila de la posición actual
    # @param col Columna de la posición actual
    # @param direction Dirección del movimiento (una de las constantes de Directions)
    # @return Array con la fila y columna de la nueva posición
    def dir_2_pos(row, col, direction)
      case direction
      when Directions::UP then [row - 1, col]
      when Directions::DOWN then [row + 1, col]
      when Directions::LEFT then [row, col - 1]
      when Directions::RIGHT then [row, col + 1]
      else [row, col]
      end
    end

    # ================================================
    # P3 - Métodos privados a implementar
    # ================================================

    # Núcleo de la lógica de movimiento. Mueve un jugador de (old) a (new)
    # @param old_row
    # @param old_col
    # @param row
    # @param col
    # @param player
    # @return Monstruo en la nueva casilla, o nil
    def put_player_2d(old_row, old_col, row, col, player)
      output = nil # Diagrama: output = null

      # Diagrama opt: [canStepOn(row, col)]
      if can_step_on?(row, col)
        # Diagrama opt: [posOK(oldRow, oldCol)]
        if pos_ok?(old_row, old_col)
          p = @players[old_row][old_col] # Diagrama 1.1: get(oldRow, oldCol)

          # Diagrama opt: [p == player]
          if p == player
            update_old_pos(old_row, old_col) # Diagrama 1.2: updateOldPos(oldRow, oldCol)

            @players[old_row][old_col] = nil # Diagrama 1.3: set(oldRow, oldCol, null)
          end
        end

        monster_pos = monster_pos(row, col) # Diagrama 1.4: monsterPos(row, col)

        # Diagrama alt: [monsterPos]
        if monster_pos
          @labyrinth[row][col] = COMBAT_CHAR # Diagrama 1.5: set(row,col, COMBAT_CHAR)

          output = @monsters[row][col] # Diagrama 1.6: get(row, col)
        else
          number = player.number # Diagrama 1.7: getNumber()

          @labyrinth[row][col] = number # Diagrama 1.8: set(row, col, number)
        end

        @players[row][col] = player # Diagrama 1.9: set(row, col, player)

        player.set_pos(row, col) # Diagrama 1.10: setPos(row, col)
      end

      output # Diagrama: return output
    end
  end
end

